### PR TITLE
Vim syntax highlighting improvements

### PR DESCRIPTION
**Pull request by compnerd**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: utils/vim/syntax/swift.vim [hunk 1]
```diff
@@ -64,20 +65,50 @@ syn keyword swiftIdentifierKeyword
...
+      \ class
+      \ enum
+      \ extension
+      \ protocol
+      \ struct
+      \ typealias
 
-syn region swiftTypeParameters start="<" end=">" contained
+syn keyword swiftNew skipwhite nextgroup=swiftTypeName
+      \ new
```
Comment (gribozavr) [inline utils/vim/syntax/swift.vim]: I think `new` is historic.  Please feel free to drop it.

File: utils/vim/syntax/swift.vim [hunk 2]
```diff
@@ -101,8 +101,13 @@ syn region swiftType contained contains=swiftType,swiftParamDelim
...
       \ /,/
-
-syn region swiftTypeParameters start="<" end=">" contained
+" <Generic Clause> (generics)
+syn region swiftTypeParameters contained contains=swiftArchetype,swiftConstraint
+      \ start="<" end=">"
+syn match swiftArchetype contained skipwhite nextgroup=swiftColonType
+      \ /\<[A-Za-z_][A-Za-z_0-9]*\>/
+syn keyword swiftConstraint contained
+      \ where
```
Comment (gribozavr) [inline utils/vim/syntax/swift.vim]: Note that constraints are being moved to the end of the declaration: https://github.com/apple/swift-evolution/blob/master/proposals/0081-move-where-expression.md

**Comment by gribozavr** [comment]
Bug report: try opening `stdlib/private/StdlibCollectionUnittest/CheckRangeReplaceableCollectionType.swift`.  The array literals (around line 300) are highlighted as types.

**Comment by compnerd** [comment]
Hmm, I will look into the issue you pointed out.  The immediate idea that I have works for that but breaks the parameter list in function declarations :-(.  lll also drop new.

**Pull request merged**